File: core/src/main/java/org/strykeforce/thirdcoast/telemetry/item/Item.java
package org.strykeforce.thirdcoast.telemetry.item;

import com.squareup.moshi.JsonWriter;
import java.io.IOException;
import java.util.Set;
import java.util.function.DoubleSupplier;
import org.strykeforce.thirdcoast.telemetry.grapher.Measure;

/**
 * An item that can be graphed. These are used as {@code Set} elements and implementing classes
 * should implement an identity-based version of equals and hashCode.
 *
 * The abstract base class implementing {@code Item} implements {@code Comparable} by comparing the
 * results returned by {@code Item#id()}.
 */
public interface Item extends Comparable<Item> {

  /**
   * Returns the underlying device id, for example, CAN bus address or PWM port.
   *
   * @return the device id.
   */
  int id();

  /**
   * A {@code String}  representing the underlying device type.
   *
   * @return the type.
   */
  String type();

  /**
   * The description of this item.
   *
   * @return the description.
   */
  String description();

  /**
   * {@code Set} of {@code Measure} items applicable to this item type.
   *
   * @return Set of Measure enumeration values.
   */
  Set<Measure> measures();

  /**
   * Suppliers that implement the measures.
   *
   * @param measure the Measure to supply.
   * @return the supplier that gives the current measurement.
   */
  DoubleSupplier measurementFor(Measure measure);

  /**
   * Provides a detailed JSON representation of the underlying device.
   *
   * @param writer the writer to write to.
   * @throws IOException if an IO error occurs.
   */
  void toJson(JsonWriter writer) throws IOException;
}

File: telemetry/src/main/kotlin/org/strykeforce/thirdcoast/telemetry/item/Item.kt
package org.strykeforce.thirdcoast.telemetry.item

import org.strykeforce.thirdcoast.telemetry.grapher.Measure
import java.util.function.DoubleSupplier

/**
 * An item that can be graphed. These are used as `Set` elements and implementing classes
 * should implement an identity-based version of equals and hashCode.
 *
 *
 * The abstract base class implementing `Item` implements `Comparable` by comparing
 * the results returned by `Item#deviceId()`.
 */
interface Item : Comparable<Item> {

    /**
     * Returns the underlying device id, for example, CAN bus address or PWM port.
     */
    val deviceId: Int

    /**
     * A `String` representing the underlying device type.
     */
    val type: String

    /**
     * The description of this item.
     */
    val description: String

    /**
     * `Set` of `Measure` items applicable to this item type.
     */
    val measures: Set<Measure>

    /**
     * Suppliers that implement the measures.
     * @return the supplier that gives the current measurement.
     */
    fun measurementFor(measure: Measure): DoubleSupplier

    override fun compareTo(other: Item): Int {
        val result = type.compareTo(other.type)
        return if (result != 0) result else deviceId.compareTo(other.deviceId)
    }

}

internal fun Boolean.toDouble() = if (this) 1.0 else 0.0